python/revng/cli/_commands/idb_converter.py

```diff
@@ -0,0 +1,655 @@
+idb_procname_to_revng_arch = {
+    "metapc": m.Architecture.x86_64,
+    "ARM": m.Architecture.arm,
+    "mips": m.Architecture.mips,
+    "mipsb": m.Architecture.mipsel,
+}
+
+revng_arch_to_metaaddr_code_type = {
+    m.Architecture.x86: MetaAddressType.Code_x86,
+    m.Architecture.x86_64: MetaAddressType.Code_x86_64,
+    m.Architecture.arm: MetaAddressType.Code_arm,
+    m.Architecture.aarch64: MetaAddressType.Code_aarch64,
+    m.Architecture.mips: MetaAddressType.Code_mips,
+    m.Architecture.mipsel: MetaAddressType.Code_mipsel,
+    m.Architecture.systemz: MetaAddressType.Code_systemz,
+}
+
+revng_arch_to_abi = {
+    m.Architecture.arm: m.ABI.Invalid,
+    m.Architecture.aarch64: m.ABI.Invalid,
+    m.Architecture.x86: m.ABI.SystemV_x86,
+    m.Architecture.x86_64: m.ABI.SystemV_x86_64,
+    m.Architecture.mips: m.ABI.SystemV_MIPS_o32,
+    m.Architecture.mipsel: m.ABI.Invalid,
+    m.Architecture.systemz: m.ABI.SystemZ_s390x,
+}
+
+revng_arch_to_abiname = {
+    m.Architecture.arm: "Invalid",
+    m.Architecture.aarch64: "Invalid",
+    m.Architecture.x86: "SystemV_x86",
+    m.Architecture.x86_64: "SystemV_x86_64",
+    m.Architecture.mips: "SystemV_MIPS_o32",
+    m.Architecture.mipsel: "Invalid",
+    m.Architecture.systemz: "SystemZ_s390x",
+}
+
+CONST_QUALIFIER = m.Qualifier(Kind=m.QualifierKind.Const)
+
+
+class IDBConverter:
+    def __init__(self, input_idb: idb.fileformat.IDB, base_addr):
+        self.idb: idb.fileformat.IDB = input_idb
+        self.api = idb.IDAPython(self.idb)
+
+        self.arch: m.Architecture = self._get_arch()
+        self.is64bit: bool = self._is_64_bit()
+        self.segments: List[m.Segment] = []
+        self.revng_types_by_id: Dict[int, RevngTypes] = {}
+        self.idb_types_to_revng_types: Dict[int, m.QualifiedType] = {}
+        self.functions: Set[m.Function] = set()
+        self.dynamic_functions: List[m.DynamicFunction] = []
+        self.imported_libraries: List[str] = []
+        self.base_addr = base_addr
+
+        self._structs_to_fixup: Set[Tuple[m.StructType, idb.typeinf.TInfo]] = set()
+        self._unions_to_fixup: Set[Tuple[m.UnionType, idb.typeinf.TInfo]] = set()
+
+        self._import_types()
+        self._fixup_structs()
+        self._fixup_unions()
+        self._import_functions()
+        self._collect_imports()
+        self._fixup_void_cabifunctiontype_arguments()
+
+    def _import_types(self):
+        """Imports initial types from the IDB. The types will be incomplete and need to be fixed"""
+        til = self.idb.til
+        for type_definition in til.types.defs:
+            assert isinstance(type_definition, idb.typeinf.TILTypeInfo)
+            the_type = type_definition.type
+            assert isinstance(the_type, idb.typeinf.TInfo)
+            self._convert_idb_type_to_revng_type(the_type, ordinal=type_definition.ordinal)
+
+    def _import_functions(self):
+        metaaddr_type = revng_arch_to_metaaddr_code_type[self.arch]
```

Comment:
We should distinguish regular ARM from Thumb here.

Reply:
Fixed.